includes/enrolment/class-sensei-enrolment-job-scheduler.php

```diff
@@ -60,23 +60,49 @@ public function init() {
 	 *
 	 * @param int  $course_id        Course post ID.
 	 * @param bool $invalidated_only Recalculate just the results that have been invalidated (set to an empty string).
-	 * @param int  $batch_size       Batch size for the job. Null will use default batch size set by job handler.
 	 *
 	 * @return Sensei_Enrolment_Course_Calculation_Job Job object.
 	 */
-	public function start_course_calculation_job( $course_id, $invalidated_only, $batch_size = null ) {
+	public function start_course_calculation_job( $course_id, $invalidated_only ) {
 		$args = [
 			'course_id'        => $course_id,
 			'invalidated_only' => $invalidated_only,
-			'batch_size'       => $batch_size,
 		];
 
+		// Make sure all previous jobs for this course are stopped.
+		$this->cancel_course_calculation_job( $course_id );
+
 		$job = new Sensei_Enrolment_Course_Calculation_Job( $args );
+		$job->start();
+
 		Sensei_Scheduler::instance()->schedule_job( $job );
 
 		return $job;
 	}
 
+	/**
+	 * Cancel all variations of a course calculation job.
+	 *
+	 * @param int $course_id Course post ID.
+	 */
+	private function cancel_course_calculation_job( $course_id ) {
+		$invalidated_only_variations = [ true, false ];
+
+		foreach ( $invalidated_only_variations as $invalidated_only ) {
+			$args = [
+				'course_id'        => $course_id,
+				'invalidated_only' => $invalidated_only,
+			];
+			$job  = new Sensei_Enrolment_Course_Calculation_Job( $args );
+
+			// If we are able to resume a current job, cancel it.
+			if ( $job->resume() ) {
+				$job->end();
+				Sensei_Scheduler::instance()->cancel_scheduled_job( $job );
```

Comment:
If there is a course recalculation job which runs for all users and we register a new job which runs for enrolled users only, does this mean that the recalculation for the pending not enrolled users will never happen?

I this is possible, I would consider removing the enrolled/unenrolled distinction on this job to avoid the complexity that it introduces. I am thinking that removing a product from a course might be rare so the performance gain might be small.

Reply:
Good catch. Do you mean the invalidated distinction? Yeah, I could do. Right now it is just the removing product (although we could also do it when switching from `publish` to `draft`). I'm wondering more for future uses of this it seems useful. Let me see how complex it is to fix this issue and if it is too complex, I'll remove it.

Reply:
Yeah, it was quite complex to fix. I removed the distinction in 0293c210f1754789602cd570677e428798a93a1b. I'll create a corresponding PR for WCPC.

Reply:
It was too complex to fix and keep working right. I've removed the distinction and created this corresponding WCPC PR https://github.com/Automattic/sensei-wc-paid-courses/pull/474

f8364a655f4939351aa31d9325ed5dbe83732a54